fix(store): remove afterEach hook in mock store

cedricduffournet opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Closes #3243

## What is the new behavior?

## Does this PR introduce a breaking change?

```
[x] Yes
[] No
```
`MockStore` will not reset all of the mocked selectors after each test. User will need to call the` MockStore.resetSelectors()` method manually in a `afterEach` hook of a `describe` block

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

cedricduffournet commented:
don't know why ci/circleci: schematics-core-check failed on CircleCI

brandonroberts commented on modules/store/testing/src/mock_store.ts:
```diff
@@ -18,9 +18,13 @@ import { MOCK_SELECTORS } from './tokens';
-      const mockStore: MockStore | undefined = TestBed.inject(MockStore);
-      if (mockStore) {
-        mockStore.resetSelectors();
+      const TestBed = getTestBed() as any;
```
Add a comment here if why this is needed, so we'll have some context later when we can potentially remove it.

cedricduffournet replied:
Ok, not sur if my english is good enough, but I tried ...

timdeschryver commented on modules/store/testing/src/mock_store.ts:
```diff
@@ -18,9 +18,21 @@ import { MOCK_SELECTORS } from './tokens';
+       */
+      const TestBed = getTestBed() as any;
+      const shouldResetSelectors = !TestBed.shouldTearDownTestingModule();
+      if (shouldResetSelectors) {
```
I think that because we don't reset the selectors anymore, that the selectors keep their "old" value.
What's the consequence of doing this? Should we provide another reset strategy?

brandonroberts replied:
We reset the selectors every time you instantiate the mock store anyway, so I'm wondering about when we'd actually need this also.

cedricduffournet replied:
Seems to be related to this: 
issue: https://github.com/ngrx/platform/issues/2044
PR: https://github.com/ngrx/platform/pull/2115

And from documentation:

> Note: MockStore will reset all of the mocked selectors after each test (in the afterEach() hook) by calling the MockStore.resetSelectors() method.

cedricduffournet replied:
executing mock_store.spec.ts pointed me out issue related to selectors not rese  ( not sur why all test pass in CI)

timdeschryver replied:
@brandonroberts we only reset the selectors that are included in a mock store instance.
With the previous implementation, that instance would be the same in the `afterEach` hook, meaning that we were we able to reset the selectors here.
With the current behavior, I don't think that `resetSelectors` is invoked (with teardown enabled), and thus that selectors won't be cleaned up in between tests.
For example, these tests are passing, while I think they should fail (and I think that these are failing on the previous version).

```ts
  it('set selectors', () => {
    store.overrideSelector(selectCollectionBookIds, [2, 3, 4] as any);
    expect(selectCollectionBookIds({})).toEqual([2, 3, 4] as any);
  });

  it('reset selectors', () => {
    // I would expect this to fail, but the selector doesn't reset in between tests
    expect(selectCollectionBookIds({})).toEqual([2, 3, 4] as any);
  });
```

timdeschryver replied:
Can we can remove this hook and tackle it if it becomes a problem?
A possible solution to keep this functionality is storing the mocked selectors in a global variable, this will be the simplest solution I think.

cedricduffournet replied:
Yes, I'm not sure what you mean and how to handle this. Should I close this PR and let you fix this ?

timdeschryver replied:
@cedricduffournet you can leave the PR open, I'm suggesting to remove the `afterEach` hook until it becomes a problem and then we can think of a solution that we're happy with. 
If @brandonroberts approves, you can remove that hook in this PR.

brandonroberts replied:
I'm good with that

timdeschryver commented on modules/store/testing/spec/mock_store.spec.ts:
```diff
@@ -541,81 +541,4 @@ describe('Cleans up after each test', () => {
     });
   });
-
-  it('should return the real value', (done: any) => {
-    TestBed.configureTestingModule({
-      imports: [
-        StoreModule.forRoot({} as any, {
-          initialState: {
-            value: 300,
-          },
-        }),
-      ],
-    });
-
-    const store = TestBed.inject(Store);
-    store.pipe(select(selectData)).subscribe((v) => {
-      expect(v).toBe(300);
-      done();
-    });
-  });
-});
```
What's the reason that this test is removed?
I think it can be re-added and that it should pass.

cedricduffournet replied:
I think this is the reason why this hook has been introduced (more info : https://github.com/ngrx/platform/issues/2044 and https://github.com/ngrx/platform/pull/2115)
If selector are not reset, this test fail because selectData return 200, which is the value set in previous test ...

timdeschryver replied:
Oh I didn't notice that, and you're right.
If that's the case, then I think we can completely remove the `Cleans up after each test` block.

cedricduffournet replied:
@timdeschryver I removed useless test.
I wonder if this PR does not introduce a breaking change. Some test that passed before will end up in error.

timdeschryver replied:
Correct, this is indeed a breaking change.
But leaving it as is will also be a breaking change.
If I'm not mistaken, the suggested implementation has the same outcome removing this code.

That's why I'm suggesting to just remove it, which will already help projects and think of a good solution if someone bumps against this problem. I'm thinking that in the majority of the cases that this won't be a problem, I'm assuming that if you're mocking selectors in a test, that you're also mocking those selectors in the following test.

timdeschryver replied:
Update: I verified that the same tests are failing with the suggested implementation:

```ts
      const TestBed = getTestBed() as any;
      const shouldResetSelectors = !TestBed.shouldTearDownTestingModule();
      if (shouldResetSelectors) {
        const mockStore: MockStore | undefined = TestBed.inject(MockStore);
        if (mockStore) {
          mockStore.resetSelectors();
        }
      }
```

cedricduffournet replied:
Yes, I had the the problem with the previous code, and removing the hook seems for me as well the best solution for now. 
Thanks for you help

timdeschryver replied:
No, thank you! 🤗

timdeschryver reviewed (commented):
I left one question, the rest of the PR seems good to me.

cedricduffournet commented:
@timdeschryver maybe we should remove or update the following text in the doc (https://ngrx.io/guide/store/testing) :

> Note: MockStore will reset all of the mocked selectors after each test (in the afterEach() hook) by calling the MockStore.resetSelectors() method.

timdeschryver commented:
@cedricduffournet good catch!

cedricduffournet commented:
@timdeschryver I updated the doc and added a test, maybe not necessary,  to show how reset selectors in afterEach hook. Let me know if you prefer that I remove it

timdeschryver commented on projects/ngrx.io/content/guide/store/testing.md:
```diff
@@ -75,7 +75,7 @@ In this example based on the [walkthrough](guide/store/walkthrough), we mock the
 <div class="alert is-helpful">
 
-**Note:** `MockStore` will reset all of the mocked selectors after each test (in the `afterEach()` hook) by calling the `MockStore.resetSelectors()` method.
+**Note:** `MockStore` will **not** reset mocked selectors after each test. You can reset selectors by calling the `MockStore.resetSelectors()` method in `afterEach()` hook.
```
I think we can remove this alert, and show how you can reset the store manually after each test.
About the re-added tests. I think we're already testing that  `resetSelector` does what it's supposed to do, so these are redundant and can be removed. 
Thoughts?

cedricduffournet replied:
Ok, that sounds good to me. I was not sure about this test.
I'll add a code example in the doc

timdeschryver commented on projects/ngrx.io/content/guide/store/testing.md:
```diff
@@ -73,11 +73,9 @@ Usage:
-<div class="alert is-helpful">
-
-**Note:** `MockStore` will reset all of the mocked selectors after each test (in the `afterEach()` hook) by calling the `MockStore.resetSelectors()` method.
+You can reset selectors by calling the `MockStore.resetSelectors()` method in `afterEach()` hook.
```
```suggestion
You can reset selectors by calling the `MockStore.resetSelectors()` method in the `afterEach()` hook.
```

timdeschryver commented on projects/ngrx.io/content/examples/store-walkthrough/src/app/tests/app.component.1.spec.ts:
```diff
@@ -96,3 +97,93 @@ describe('AppComponent', () => {
+    });
+  });
+
+  it('should return the real value', (done: any) => {
+    TestBed.configureTestingModule({
+      imports: [
+        StoreModule.forRoot({} as any, {
+          initialState: {
+            books: [
+              {
+                id: 'realId',
+                volumeInfo: {
+                  title: 'Real Title',
+                  authors: ['Real Author'],
+                },
+              },
+            ],
+          },
+        }),
+      ],
+    });
+
+    const store = TestBed.inject(Store);
+
+    store.select(selectBooks).subscribe((mockBooks) => {
+      expect(mockBooks).toEqual([
+        {
+          id: 'realId',
+          volumeInfo: {
+            title: 'Real Title',
+            authors: ['Real Author'],
+          },
+        },
+      ]);
+      done();
+    });
+  });
```
I think it would be better to leave this out, to keep the example concise.
```suggestion
```

timdeschryver commented on projects/ngrx.io/content/examples/store-walkthrough/src/app/tests/app.component.1.spec.ts:
```diff
@@ -96,3 +97,93 @@ describe('AppComponent', () => {
+
+//#docregion resetMockSelector
+describe('AppComponent reset selectors', () => {
+  let store: MockStore<AppState>;
```
Just like `Store`, we want to remove the generic.

```suggestion
  let store: MockStore;
```

timdeschryver commented on projects/ngrx.io/content/examples/store-walkthrough/src/app/tests/app.component.1.spec.ts:
```diff
@@ -96,3 +97,93 @@ describe('AppComponent', () => {
+    TestBed.configureTestingModule({
+      providers: [
+        provideMockStore({
+          initialState: [
+            {
+              id: 'firstId',
+              volumeInfo: {
+                title: 'Initial Title',
+                authors: ['Initial Author'],
+              },
+            },
+          ],
```
When mocking selectors, the state doesn't matter.
So, I would prefer to leave it out.

```suggestion
```

timdeschryver commented on projects/ngrx.io/content/examples/store-walkthrough/src/app/tests/app.component.1.spec.ts:
```diff
@@ -96,3 +97,93 @@ describe('AppComponent', () => {
+        }),
+      ],
+    });
+
```
We need to grab an instance of `store` here.

```suggestion

const store = TestBed.inject(MockStore);

```

timdeschryver commented on projects/ngrx.io/content/examples/store-walkthrough/src/app/tests/app.component.1.spec.ts:
```diff
@@ -96,3 +97,93 @@ describe('AppComponent', () => {
+  let store: MockStore<AppState>;
+
+  afterEach(() => {
+    store.resetSelectors();
```
Just for safety.

```suggestion
    store?.resetSelectors();
```

timdeschryver reviewed (commented):
Thanks again @cedricduffournet .
I left a few comments to make the example more compact.

timdeschryver reviewed (approved):
Thanks for your patience and thoroughness @cedricduffournet .
